### ISSUE TITLE
Codec cloning with multiple workers seems to break codecs that keep internal state

### ISSUE THREAD

**Issue by jorritfolmer** [logstash-plugins/logstash-input-udp#36]
Over at logstash-codec-netflow I'm seeing data loss issues when using udp plugin version >3.2.0

Copy of the reproducer text at https://github.com/logstash-plugins/logstash-codec-netflow/issues/122#issuecomment-375959049:

To reproduce: 
1) Configure 2 or more udp input workers
2) Logstash-Input-UDP >= 3.2.0
3) Conf:
    ```
    input {
      udp {
        port                 => 2055
        receive_buffer_bytes => 16777216
        codec                => netflow
        workers              => 2
      }
    }
    ``` 
4) Run [netflow_bench_cisco_asa.py](https://github.com/logstash-plugins/logstash-codec-netflow/blob/master/spec/codecs/benchmarks/netflow_bench_cisco_asa.py) against Logstash

Expected results: 
- all 210000 flows decoded (14 flows per packet x 15000 packets)

Actual results: 
- only some percentage of flows decoded
- lots of "no template received" warnings.

Discussion: With Netflow-Input-UDP >= 3.2.0 Netflow data packets can only be decoded by the worker that has previously received a template to decode with. Netflow_bench_cisco_asa.py sends only 1 template before sending 15000 packets. Since the codec is now cloned for each worker with this commit in https://github.com/logstash-plugins/logstash-input-udp/commit/ecf93e1a82f820fef7166517aa2926e206d8abfb,  templates are no longer shared between workers. This means only a percentage of flows can be decoded, and others not.

**Comment by jsvd** [comment]
@jorritfolmer this is bad indeed, we didn't take into account how the netflow codec maintains state. Other plugins do it per channel/connection, like json_lines and line. These keep state for each producer, while netflow's pretty much global.

I see a few paths forward:
a) marking all plugin releases that perform codec cloning as broken due to being breaking changes in a minor release, resulting in yanking some versions of udp/http/tcp
b) patching netflow codec to prevent cloning (e.g. `def clone; self; end`), this ensures the same instance is used across workers
c) adding a property to codecs to know if they are threadsafe or not. this would allow plugins to chose when to reuse an instance vs clone the plugin

@colinsurprenant thoughts on this?

**Comment by jorritfolmer** [comment]
B) works, just tested it.

**Comment by colinsurprenant** [comment]
so, yeah, this is a bit nasty - just discussed with @jsvd and we agree that as a short term solution B) is good and will work with all published udp input versions. we will discuss on how to correctly tackle this moving forward. 
Thanks for the heads up @jorritfolmer

**Comment by jorritfolmer** [comment]
Ok thanks I've included the fix in 3.11.4 of logstash-codec-netflow

**Comment by jsvd** [comment]
Many thanks for the change @jorritfolmer, apologies for unexpected breakage.
